ecs_pod_creator.go

```diff
@@ -236,6 +326,253 @@ func (e *EnvironmentVariable) SetSecretOptions(opts SecretOptions) *EnvironmentV
 	return e
 }
 
+// Validate checks that the environment variable name is given and that either
+// the raw environment variable value or the referenced secret is given.
+func (e *EnvironmentVariable) Validate() error {
+	catcher := grip.NewBasicCatcher()
+	catcher.NewWhen(e.Name == nil, "must specify a name")
+	catcher.NewWhen(e.Value == nil && e.SecretOpts == nil, "must either specify a value or reference a secret")
+	catcher.NewWhen(e.Value != nil && e.SecretOpts != nil, "cannot both specify a value and reference a secret")
+	if e.SecretOpts != nil {
+		catcher.Wrap(e.SecretOpts.Validate(), "invalid secret options")
+	}
+	return catcher.Resolve()
+}
+
+// SecretOptions represents a secret with a name and value that may or may not
+// be owned by its pod.
+type SecretOptions struct {
+	PodSecret
+	// Exists determines whether or not the secret already exists or must be
+	// created before it can be used.
+	Exists *bool
+}
+
+// NewSecretOptions returns new uninitialized options for a secret.
+func NewSecretOptions() *SecretOptions {
+	return &SecretOptions{}
+}
+
+// SetName sets the secret name.
+func (s *SecretOptions) SetName(name string) *SecretOptions {
+	s.Name = &name
+	return s
+}
+
+// SetValue sets the secret value.
+func (s *SecretOptions) SetValue(val string) *SecretOptions {
+	s.Value = &val
+	return s
+}
+
+// SetExists sets whether or not the secret already exists or or must be
+// created.
+func (s *SecretOptions) SetExists(exists bool) *SecretOptions {
+	s.Exists = &exists
+	return s
+}
+
+// Validate validates that the secret name is given and that either the secret
+// already exists or the new secret's value is given
+func (s *SecretOptions) Validate() error {
+	catcher := grip.NewBasicCatcher()
+	catcher.NewWhen(s.Name == nil, "must specify a name")
+	catcher.NewWhen(!utility.FromBoolPtr(s.Exists) && s.Value == nil, "either a new secret's value must be given or the secret must already exist")
```

Comment:
Can the value be an empty string?

Reply:
Same thing as the other comment about `NamedSecret` - little weird, but not wrong.